File: go/bit/137_single_number.go
package bit

// 只出现一次的数字 II
// 除某个元素仅出现 一次 外，其余每个元素都恰出现 三次 。请你找出并返回那个只出现了一次的元素。
func singleNumber2(nums []int) int {
	// 如果每个元素出现了三次，那么他的每位二进制bit位 也出现了三次
	// 所以统计每个元素的二进制位的个数和，然后对3求模，得到二进制表示转为十进制即是结果

	var res int32

	// nums 中的元素是32位数字，所以遍历32次，计算每次最后以为
	for i := 0; i < 32; i++ {
		// 初始没位置位0
		cnt := int32(0)

		for _, v := range nums {
			// >> i 位，右移i位，意味着计算到元素v 的第i位的数字了
			// &1 取出该数字
			cnt += int32(v) >> int32(i) & 1
		}
		// 将计算的cnt 左移 i 位，然后赋值到res 的i 位
		res |= cnt % 3 << i
	}

	return int(res)
}

File: rust/src/bit/_137_single_number.rs
// 除某个元素仅出现 一次 外，其余每个元素都恰出现 三次 。请你找出并返回那个只出现了一次的元素。
pub fn single_number(nums: Vec<i32>) -> i32 {
    // 如果某个元素出现了三次，那么将其二进制bit位，按位想加，则其和应该是3的倍数
    // 思路: 1. 按位想加
    // 2. 对3 取模
    // 3. 得到的元素就是出现一次的元素的二进制bnui
    let mut res = 0;

    // 32 位元素，循环32次，计算按位想加的和
    for i in 0..32 {
        let mut sum = 0;

        for v in nums.iter() {
            // v &1 ： 获取 v 的最低位（最右面bit 位的值）
            // 第 i 次计算，需要将 v >> i ， 右移i位，获取第i位的值
            sum += (v >> i) & 1
        }
        // 将sum 取模的值，左移i位，放到res i 位的位置上
        res |= (sum % 3) << i
    }

    res
}